File: android/Streamo/app/src/main/java/com/streamo/app/util/Format.kt
package com.streamo.app.util

object Format {

    fun time(seconds: Double): String {
        if (!seconds.isFinite() || seconds <= 0) return "0:00"
        val total = seconds.toInt()
        val h = total / 3600
        val m = (total % 3600) / 60
        val s = total % 60
        return if (h > 0) {
            String.format("%d:%02d:%02d", h, m, s)
        } else {
            String.format("%d:%02d", m, s)
        }
    }

    fun percent(position: Double, duration: Double): Double {
        if (duration <= 0) return 0.0
        return (position / duration * 100).coerceIn(0.0, 100.0)
    }

    fun viewedMinutes(position: Double?): String? {
        val p = position ?: return null
        if (p <= 0) return null
        val minutes = (p / 60).toInt().coerceAtLeast(1)
        return if (minutes == 1) "Visto 1 min" else "Visti $minutes min"
    }

    fun watchTime(seconds: Double): String {
        val totalMin = (seconds / 60).toInt()
        val h = totalMin / 60
        val m = totalMin % 60
        return if (h > 0) {
            if (m > 0) "$h h $m min" else "$h h"
        } else {
            "$m min"
        }
    }
}

File: vidaaos/src/util/format.ts
// Port of Format.kt. Time/percent/minutes/watch-time formatting.
// Output strings match the Kotlin source (Italian where the source is Italian).

export function time(seconds: number): string {
  if (!Number.isFinite(seconds) || seconds <= 0) return '0:00';
  const total = Math.floor(seconds);
  const h = Math.floor(total / 3600);
  const m = Math.floor((total % 3600) / 60);
  const s = total % 60;
  return h > 0 ? `${h}:${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}` : `${m}:${String(s).padStart(2, '0')}`;
}

export function percent(position: number, duration: number): number {
  if (duration <= 0) return 0.0;
  return Math.min(100.0, Math.max(0.0, (position / duration) * 100));
}

export function viewedMinutes(position: number | null | undefined): string | null {
  if (position == null) return null;
  if (position <= 0) return null;
  const minutes = Math.max(1, Math.floor(position / 60));
  return minutes === 1 ? 'Visto 1 min' : `Visti ${minutes} min`;
}

export function watchTime(seconds: number): string {
  const totalMin = Math.floor(seconds / 60);
  const h = Math.floor(totalMin / 60);
  const m = totalMin % 60;
  if (h > 0) return m > 0 ? `${h} h ${m} min` : `${h} h`;
  return `${m} min`;
}

// "Xh Ymin rimasti" — remaining-time label used by the player/UI. Not in Format.kt
// directly but derived from watchTime-style formatting across the app; kept here as
// the single Italian remaining-time formatter.
export function remainingTime(secondsRemaining: number): string {
  if (!Number.isFinite(secondsRemaining) || secondsRemaining <= 0) return '0 min rimasti';
  const totalMin = Math.floor(secondsRemaining / 60);
  const h = Math.floor(totalMin / 60);
  const m = totalMin % 60;
  if (h > 0) return m > 0 ? `${h}h ${m}min rimasti` : `${h}h rimasti`;
  return `${m}min rimasti`;
}